sdl.cpp

```diff
@@ -0,0 +1,294 @@
+#include <string>
+#include <SDL.h>
+#include <SDL_image.h>
+#include <SWI-cpp2.h>
+
+PREDICATE(sdl_init_, 1)
+{
+  if (SDL_Init(A1.as_uint32_t()) < 0)
+  {
+    throw PlException(PlTerm_string(SDL_GetError()));
+  }
+  return true;
+}
+
+PREDICATE(sdl_quit, 0)
+{
+  SDL_Quit();
+  return true;
+}
+
+PREDICATE(sdl_createwindow_, 7)
+{
+  SDL_Window *window = SDL_CreateWindow(A2.as_string().c_str(),
+                                        A3.as_int(), A4.as_int(),
+                                        A5.as_int(), A6.as_int(),
+                                        A7.as_uint32_t());
+  if (window == NULL)
+  {
+    throw PlException(PlTerm_string(SDL_GetError()));
+  }
+  return A1.unify_pointer(window);
+}
+
+PREDICATE(sdl_destroywindow, 1)
+{
+  SDL_DestroyWindow((SDL_Window *) A1.as_pointer());
+  return true;
+}
+
+PREDICATE(sdl_createrenderer_, 4)
+{
+  SDL_Renderer *renderer = SDL_CreateRenderer((SDL_Window *) A2.as_pointer(),
+                                              A3.as_int(),
+                                              A4.as_uint32_t());
+  if (renderer == NULL)
+  {
+    throw PlException(PlTerm_string(SDL_GetError()));
+  }
+  return A1.unify_pointer(renderer);
+}
+
+PREDICATE(sdl_destroyrenderer, 1)
+{
+  SDL_DestroyRenderer((SDL_Renderer *) A1.as_pointer());
+  return true;
+}
+
+PREDICATE(sdl_renderclear, 1)
+{
+  if (SDL_RenderClear((SDL_Renderer *) A1.as_pointer()) < 0)
+  {
+    throw PlException(PlTerm_string(SDL_GetError()));
+  }
+  return true;
+}
+
+const SDL_Rect
+get_rect(PlTerm term)
+{
+  const SDL_Rect rect = {
+    term[1].as_int(), term[2].as_int(), term[3].as_int(),
+    term[4].as_int()
+  };
+  return rect;
+}
+
+PREDICATE(sdl_rendercopy_, 4)
+{
+  SDL_Rect *srcrect_p = NULL;
+  SDL_Rect srcrect;
+  if (A3.is_compound())
+  {
+    srcrect = get_rect(A3);
+    srcrect_p = &srcrect;
+  }
+  SDL_Rect *dstrect_p = NULL;
+  SDL_Rect dstrect;
+  if (A4.is_compound())
+  {
+    dstrect = get_rect(A4);
+    dstrect_p = &dstrect;
+  }
+  if (SDL_RenderCopy((SDL_Renderer *) A1.as_pointer(),
+                     (SDL_Texture *) A2.as_pointer(),
+                     srcrect_p, dstrect_p) < 0)
+  {
+    throw PlException(PlTerm_string(SDL_GetError()));
+  }
+  return true;
+}
+
+PREDICATE(sdl_renderpresent, 1)
+{
+  SDL_RenderPresent((SDL_Renderer *) A1.as_pointer());
```

Comment:
`static_cast<SDL_Renderer*>(A1.as_pointer())`

Reply:
done in [d45f50d](https://github.com/kwon-young/sdlpl/pull/1/commits/d45f50dfd0215a47e986f2453bcc13aa0570eec3)